### ISSUE TITLE
Improve math.isqrt() emulation

### ISSUE THREAD

**Issue by rhettinger** [more-itertools/more-itertools#732]
Currently, `math.isqrt()` is emulated with `lambda x: int(math.sqrt(x))`.

First, the `int()` should be `floor()` to match the intended semantics.

Second, the revised approach only works for up to 53 bits and then it gives wrong answers:

```
>>> n = 9007199326062755
>>> isqrt(n) == floor(sqrt(n))
False
```

Third, it should raise a `TypeError` for a floating point input:

```
>>> n = 3.14159
>>> floor(sqrt(n))
1
>>> isqrt(n)
Traceback (most recent call last):
...
TypeError: 'float' object cannot be interpreted as an integer
```

Fourth, the `sqrt()` overflows if the input is larger than the range of floats:

```
>>> n = 2 << 1030
>>> r = isqrt(n)
>>> floor(sqrt(n))
Traceback (most recent call last):
...
OverflowError: int too large to convert to float
```

Here is some suggested code:

```
# math.isqrt is available for Python 3.8+
try:

    _isqrt = math.isqrt

except AttributeError:

    def _isqrt(n):
        root = n >> (n.bit_length() // 2)
        if n < 0:
            raise ValueError('isqrt() argument must be nonnegative')
        if n < 0x20000000000000:
            return math.floor(math.sqrt(n))
        while not (root * root <= n < (root + 1) * (root + 1)):
            quotient = n // root
            root = (quotient + root) // 2
        return root

```

Note, the large hex constant `0x20000000000000` is just `sys.float_info.radix ** sys.float_info.mant_dig`.

**Comment by bbayles** [comment]
Thanks for the suggestions; I'll incorporate these.

**Comment by bbayles** [comment]
Since `math.isqrt` is available from Python 3.8 ([ref](https://docs.python.org/3/library/math.html#math.isqrt)), and Python 3.7 is EOL, I decided to drop Python 3.7 support and use `math.isqrt` directly.
